File: src/数组/54.螺旋数组.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix)
    {
        if (matrix.empty())
            return {};

        int rowNum = matrix.size();
        int colNum = matrix[0].size();
        int startX = 0, startY = 0;
        vector<int> res;

        while (rowNum > 0 && colNum > 0) {
            // If there's only one row/column left, add them all
            if (rowNum == 1) {
                for (int j = 0; j < colNum; j++) {
                    res.push_back(matrix[startX][startY + j]);
                }
                break;
            } else if (colNum == 1) {
                for (int i = 0; i < rowNum; i++) {
                    res.push_back(matrix[startX + i][startY]);
                }
                break;
            }

            // Top row
            for (int j = 0; j < colNum - 1; j++) {
                res.push_back(matrix[startX][startY + j]);
            }
            // Right column
            for (int i = 0; i < rowNum - 1; i++) {
                res.push_back(matrix[startX + i][startY + colNum - 1]);
            }
            // Bottom row
            for (int j = colNum - 1; j > 0; j--) {
                res.push_back(matrix[startX + rowNum - 1][startY + j]);
            }
            // Left column
            for (int i = rowNum - 1; i > 0; i--) {
                res.push_back(matrix[startX + i][startY]);
            }

            startX++;
            startY++;
            rowNum -= 2;
            colNum -= 2;
        }

        return res;
    }
};

int main()
{
    Solution s;
    vector<vector<int>> matrix = { { 6, 9, 7 } };
    vector<int> res;
    res = s.spiralOrder(matrix);
    for (auto& x : res) {
        cout << x << " ";
    }
}

File: 数组/54.螺旋数组.py
from typing import List


class Solution:

    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])
        l, r, t, b = 0, n - 1, 0, m - 1
        res = []
        if m == 1:
            return matrix[0]
        if n == 1:
            for i in range(m):
                res.append(matrix[i][0])
            return res
        while True:
            for i in range(l, r + 1):
                res.append(matrix[t][i])
            t += 1
            if t > b:
                break
            for i in range(t, b + 1):
                res.append(matrix[i][r])
            r -= 1
            if r < l:
                break
            for i in range(r, l - 1, -1):
                res.append(matrix[b][i])
            b -= 1
            if b < t:
                break
            for i in range(b, t - 1, -1):
                res.append(matrix[i][l])
            l += 1
            if l > r:
                break
        return res


if __name__ == "__main__":
    s = Solution()
    print(s.spiralOrder([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
